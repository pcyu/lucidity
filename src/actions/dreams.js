import uuid from 'uuid'

// ADD_DREAM
export const addDream = ({ title = '', entry = '', hoursSlept = 0, dreamType = '', createdAt = 0, lucidityLevel = 0 } = {}) => ({
  type: 'ADD_DREAM',
  dream: {
    id: uuid(),
    title,
    entry,
    hoursSlept,
    dreamType,
    createdAt,
    lucidityLevel
  }
})

// REMOVE_DREAM
export const removeDream = ({ id }) => ({
  type: 'REMOVE_DREAM',
  id
})

// EDIT_DREAM
export const editDream = (id, updates) => ({
  type: 'EDIT_DREAM',
  id,
  updates
})