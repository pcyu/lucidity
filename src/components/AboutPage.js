import React from 'react'

const AboutPage = () => (
  <div>
    This is my from about page
  </div>
)

export default AboutPage