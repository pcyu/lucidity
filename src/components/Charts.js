import React from 'react'
import { connect } from 'react-redux'
import { Bar, Line, Doughnut} from 'react-chartjs-2'

class Charts extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      chartData: {
        labels: ['Monday', 'Tuesday', 'Wednesday'],
        datasets: [
          {
            label: 'Days',
            data: [1, 2, 3],
            backgroundColor: ['rgb(0, 121, 211)']
          }
        ]
      }
    }
  }
  render() {
    console.log(this.props.dreams)
    return (
      <div>
        CHART COMPONENT
        <Bar
          data={this.state.chartData}
          options={{
            title: {
              display: true,
              text: 'TITLE TEXT'
            },
            legend: {
              display: true
            }
          }}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  dreams: state.dreams
})

export default connect(mapStateToProps)(Charts)