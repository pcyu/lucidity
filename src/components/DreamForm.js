import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

class DreamForm extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      title: props.dream ? props.dream.title : '',
      entry: props.dream ? props.dream.entry : '',
      hoursSlept: props.dream ? props.dream.hoursSlept.toString() : '',
      dreamType: props.dream ? props.dream.dreamType : 'normal',
      createdAt: props.dream ? moment(props.dream.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      buttonText: ''
    }
  }
  onTitleChange = (e) => {
    const title = e.target.value
    this.setState(() => ({ title }))
  }
  onEntryChange = (e) => {
    const entry = e.target.value
    this.setState(() => ({ entry }))
  }
  onHoursSleptChange = (e) => {
    const hoursSlept = e.target.value
    if (!hoursSlept || hoursSlept.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ hoursSlept }))
    }
  }
  onDreamTypeChange = (e) => {
    const dreamType = e.target.value
    this.setState(() => ({ dreamType }))
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.title || !this.state.entry || !this.state.hoursSlept) {
      this.setState(() => ({ error: 'Please fill in all fields' }))
    } else {
      this.setState(() => ({ error: ''}))
      
      this.props.onSubmit({
        title: this.state.title,
        entry: this.state.entry,
        hoursSlept: parseFloat(this.state.hoursSlept, 10),
        createdAt: this.state.createdAt.valueOf(),
        dreamType: this.state.dreamType
      })
    }
  }
  componentWillMount() {
    !!this.state.title === false ? this.state.buttonText = 'Add Dream' : this.state.buttonText = 'Edit Dream' 
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text" 
            placeholder="Title" 
            autoFocus 
            value={this.state.title}
            onChange={this.onTitleChange}
            />
          <input 
            type="text" 
            placeholder="Hours Slept"
            value={this.state.hoursSlept}
            onChange={this.onHoursSleptChange}
          />
          <select 
            value={this.state.dreamType}
            onChange={this.onDreamTypeChange}>
            <option value="normal">Normal</option>
            <option value="lucid">Lucid</option>
            <option value="nightmare">Nightmare</option>
          </select>
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea 
            placeholder="What did you dream about?" 
            value={this.state.entry}
            onChange={this.onEntryChange}>
          </textarea>
          <button>
            {this.state.buttonText}
          </button>
        </form>
      </div>
    )
  }
}

export default DreamForm