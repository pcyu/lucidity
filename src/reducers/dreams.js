// Dreams Reducer
const dreamsReducerDefaultState = []

const dreamsReducer = (state = dreamsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_DREAM':
      return [
        ...state,
        action.dream
      ]
    case 'REMOVE_DREAM':
      return state.filter(({ id }) => id !== action.id)
    case 'EDIT_DREAM':
      return state.map((dream) => {
        if (dream.id === action.id) {
          return {
            ...dream,
            ...action.updates
          }
        } else {
          return dream
        }
      })
    default:
      return state
  }
}

export default dreamsReducer