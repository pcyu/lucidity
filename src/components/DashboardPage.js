import React from 'react'
import Charts from './Charts'
import Stats from './Stats'

const DashboardPage = () => (
  <div>
    This is from my dashboard component
    <Stats />
    <Charts />
  </div>
)

export default DashboardPage