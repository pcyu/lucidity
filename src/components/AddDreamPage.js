import React from 'react'
import { connect } from 'react-redux'
import DreamForm from './DreamForm'
import { addDream } from '../actions/dreams'

export class AddDreamPage extends React.Component {
  onSubmit = (dream) => {
    this.props.addDream(dream)
    this.props.history.push('/dreams')
  }
  render() {
    return (
      <div>
        <h1>Add Dream</h1>
        <DreamForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({ 
  addDream: (dream) => dispatch(addDream(dream)) 
})

export default connect(undefined, mapDispatchToProps)(AddDreamPage)