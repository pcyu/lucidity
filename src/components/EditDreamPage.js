import React from 'react'
import { connect } from 'react-redux'
import DreamForm from './DreamForm'
import { editDream, removeDream } from '../actions/dreams'

export class EditDreamPage extends React.Component {
  onSubmit = (dream) => {
    this.props.editDream(this.props.dream.id, dream)
    this.props.history.push('/dreams')
  }
  onRemove = () => {
    this.props.removeDream(this.props.dream)
    this.props.history.push('/dreams')
  }
  render() {
    return (
      <div>
        <DreamForm
          dream={this.props.dream}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
}


const mapStateToProps = (state, props) => ({
  dream: state.dreams.find((dream) => dream.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
  editDream: (id, dream) => dispatch(editDream(id, dream)),
  removeDream: (dream) => dispatch(removeDream(dream))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditDreamPage)