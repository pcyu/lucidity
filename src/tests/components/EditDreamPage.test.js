import React from 'react'
import { shallow } from 'enzyme'
import dreams from '../fixtures/dreams'
import { EditDreamPage } from '../../components/EditDreamPage'

let editDream, removeDream, history, wrapper

beforeEach(() => {
  editDream = jest.fn()
  removeDream = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(
    <EditDreamPage 
      editDream={editDream} 
      removeDream={removeDream} 
      history={history}
      dream={dreams[2]}
    />)
})

test('should render EditDreamPage', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle editDream', () => {
  wrapper.find('DreamForm').prop('onSubmit')(dreams[2])
  expect(history.push).toHaveBeenLastCalledWith('/dreams')
  expect(editDream).toHaveBeenLastCalledWith(dreams[2].id, dreams[2])
})

test('should handle removeDream', () => {
  wrapper.find('button').simulate('click')
  expect(history.push).toHaveBeenLastCalledWith('/dreams')
  expect(removeDream).toHaveBeenLastCalledWith(dreams[2])
})