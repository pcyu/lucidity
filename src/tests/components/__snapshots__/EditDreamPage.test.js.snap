// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditDreamPage 1`] = `
<div>
  <DreamForm
    dream={
      Object {
        "createdAt": 345600000,
        "dreamType": "nightmare",
        "entry": "this is dream three entry",
        "hoursSlept": 8,
        "id": "3",
        "title": "dream three",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
